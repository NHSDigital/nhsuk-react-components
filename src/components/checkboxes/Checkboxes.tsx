import React, { HTMLProps, PureComponent } from 'react';
import classNames from 'classnames';
import { FormElementProps } from '../../util/types/FormTypes';
import FormGroup from '../../util/FormGroup';
import CheckboxContext, { ICheckboxContext } from './CheckboxContext';
import Box from './components/Box';
import { generateRandomName } from '../../util/RandomID';

interface CheckboxesProps extends HTMLProps<HTMLDivElement>, FormElementProps {
  idPrefix?: string;
}

type CheckboxesState = {
  conditionalBoxes: Array<string>;
};

class Checkboxes extends PureComponent<CheckboxesProps, CheckboxesState> {
  private boxCount = 0;

  private boxReferences: Array<string> = [];

  private boxIds: Record<string, string> = {};

  constructor(props: CheckboxesProps) {
    super(props);
    this.state = {
      conditionalBoxes: [],
    };
  }

  leaseReference = (): string => {
    const reference = generateRandomName();
    if (this.boxReferences.includes(reference)) {
      return this.leaseReference();
    }
    this.boxReferences.push(reference);
    return reference;
  };

  unleaseReference = (reference: string): void => {
    this.boxReferences = this.boxReferences.filter((ref) => ref !== reference);
  };

  setConditional = (boxReference: string, hasConditional: boolean): void => {
    this.setState((state) => {
      const currentHasConditional = state.conditionalBoxes.includes(boxReference);
      if (currentHasConditional && hasConditional === false) {
        return {
          ...state,
          conditionalBoxes: state.conditionalBoxes.filter((ref) => ref !== boxReference),
        };
      }
      if (!currentHasConditional && hasConditional === true) {
        return {
          ...state,
          conditionalBoxes: [...state.conditionalBoxes, boxReference],
        };
      }
      return state;
    });
  };

  getBoxId = (id: string, reference: string): string => {
    const { idPrefix } = this.props;
    if (reference in this.boxIds) {
      return this.boxIds[reference];
    }
    this.boxCount += 1;
    this.boxIds[reference] = `${idPrefix || id}-${this.boxCount}`;
    return this.boxIds[reference];
  };

  resetBoxIds = (): void => {
    this.boxCount = 0;
    this.boxIds = {};
  };

  static Box = Box;

  render(): JSX.Element {
    const { children, ...rest } = this.props;
    return (
      <FormGroup<CheckboxesProps> inputType="checkboxes" {...rest}>
        {/* eslint-disable-next-line @typescript-eslint/no-unused-vars */}
        {({ className, name, id, idPrefix, error, ...restRenderProps }) => {
          this.resetBoxIds();
          const containsConditional = this.state.conditionalBoxes.length > 0;
          const contextValue: ICheckboxContext = {
            name,
            getBoxId: (reference) => this.getBoxId(id, reference),
            setConditional: this.setConditional,
            leaseReference: this.leaseReference,
            unleaseReference: this.unleaseReference,
          };
          return (
            <div
              className={classNames(
                'nhsuk-checkboxes',
                { 'nhsuk-checkboxes--conditional': containsConditional },
                className,
              )}
              id={id}
              {...restRenderProps}
            >
              <CheckboxContext.Provider value={contextValue}>{children}</CheckboxContext.Provider>
            </div>
          );
        }}
      </FormGroup>
    );
  }
}

export default Checkboxes;
