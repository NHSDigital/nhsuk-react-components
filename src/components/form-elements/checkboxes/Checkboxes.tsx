'use client';

import React, { HTMLProps, useEffect } from 'react';
import classNames from 'classnames';
import { FormElementProps } from '@util/types/FormTypes';
import SingleInputFormGroup from '@components/utils/SingleInputFormGroup';
import CheckboxContext, { ICheckboxContext } from './CheckboxContext';
import Box from './components/Box';
import Divider from './components/Divider';
import { generateRandomName } from '@util/RandomID';
import CheckboxJs from '@resources/checkboxes';

interface CheckboxesProps extends HTMLProps<HTMLDivElement>, FormElementProps {
  idPrefix?: string;
}

const Checkboxes = ({ children, idPrefix, ...rest }: CheckboxesProps) => {
  const _boxReferences: string[] = [];
  let _boxCount: number = 0;
  let _boxIds: Record<string, string> = {};

  useEffect(() => {
    CheckboxJs();
  }, []);

  const getBoxId = (id: string, reference: string): string => {
    if (reference in _boxIds) {
      return _boxIds[reference];
    }
    _boxCount++;
    _boxIds[reference] = `${idPrefix ?? id}-${_boxCount}`;

    return _boxIds[reference];
  };

  const leaseReference = (): string => {
    let reference: string = '';
    do {
      reference = generateRandomName();
    } while (_boxReferences.includes(reference));

    _boxReferences.push(reference);
    return reference;
  };

  const unleaseReference = (reference: string): void => {
    _boxReferences.splice(_boxReferences.indexOf(reference), 1);
  };

  const resetCheckboxIds = (): void => {
    _boxCount = 0;
    _boxIds = {};
  };

  return (
    <SingleInputFormGroup<CheckboxesProps> inputType="checkboxes" {...rest}>
      {/* eslint-disable-next-line @typescript-eslint/no-unused-vars */}
      {({ className, name, id, idPrefix, error, ...restRenderProps }) => {
        resetCheckboxIds();
        const contextValue: ICheckboxContext = {
          name,
          getBoxId: (reference) => getBoxId(id, reference),
          leaseReference,
          unleaseReference,
        };
        return (
          <div className={classNames('nhsuk-checkboxes', className)} id={id} {...restRenderProps}>
            <CheckboxContext.Provider value={contextValue}>{children}</CheckboxContext.Provider>
          </div>
        );
      }}
    </SingleInputFormGroup>
  );
};

Checkboxes.Box = Box;
Checkboxes.Divider = Divider;

export default Checkboxes;
