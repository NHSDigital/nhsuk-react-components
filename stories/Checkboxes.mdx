import {Meta} from "@storybook/blocks";

<Meta title="Components/Checkboxes" />

# Checkboxes

This component can be found in the `nhsuk-frontend` repository <a href="https://github.com/nhsuk/nhsuk-frontend/tree/master/packages/components/checkboxes" target="_blank" rel="noopener noreferrer">here</a>.

## Implementation Notes

For details on the new form design pattern, check the documentation for the `Form` component.

The `Checkbox` component provides a `CheckboxContext` context, which the `Checkbox.Box` components use. When each box initially renders, it will attempt to assign itself an `id` by calling the `getBoxId` method in the context. This will assign each box a sequential ID using either the `idPrefix` or `name` supplied to the Checkbox. If neither are provided, the element will generate it's own unique identifier.

## Usage

### Standard

```jsx
import { Checkboxes } from "nhsuk-react-components";

const Element = () => {
    return (
        <Checkboxes name="nationality" id="nationality">
            <Checkboxes.Box value="british">British</Checkboxes.Box>
            <Checkboxes.Box value="irish">Irish</Checkboxes.Box>
            <Checkboxes.Box value="other">Citizen of another country</Checkboxes.Box>
        </Checkboxes>
    );
}
```
