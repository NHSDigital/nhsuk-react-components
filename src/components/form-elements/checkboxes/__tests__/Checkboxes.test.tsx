import React from 'react';
import { render } from '@testing-library/react';
import Checkboxes from '../';

describe('Checkboxes', () => {
  it('matches snapshot', () => {
    const { container } = render(
      <form>
        <Checkboxes id="example" name="example">
          <Checkboxes.Box value="animal">Waste from animal carcasses</Checkboxes.Box>
          <Checkboxes.Box value="mines">Waste from mines or quarries</Checkboxes.Box>
          <Checkboxes.Box value="farm">Farm or agricultural waste</Checkboxes.Box>
        </Checkboxes>
      </form>,
    );

    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with string error', () => {
    const { container } = render(
      <form>
        <Checkboxes id="example" name="example" error="Example error">
          <Checkboxes.Box value="animal">Waste from animal carcasses</Checkboxes.Box>
          <Checkboxes.Box value="mines">Waste from mines or quarries</Checkboxes.Box>
          <Checkboxes.Box value="farm">Farm or agricultural waste</Checkboxes.Box>
        </Checkboxes>
      </form>,
    );

    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with boolean error', () => {
    const { container } = render(
      <form>
        <Checkboxes id="example" name="example" error={true}>
          <Checkboxes.Box value="animal">Waste from animal carcasses</Checkboxes.Box>
          <Checkboxes.Box value="mines">Waste from mines or quarries</Checkboxes.Box>
          <Checkboxes.Box value="farm">Farm or agricultural waste</Checkboxes.Box>
        </Checkboxes>
      </form>,
    );

    expect(container).toMatchSnapshot();
  });

  it('Matches the snapshot with an exclusive checkbox', () => {
    const { container } = render(
      <form>
        <Checkboxes id="example" name="example">
          <Checkboxes.Box value="animal">Waste from animal carcasses</Checkboxes.Box>
          <Checkboxes.Box value="mines">Waste from mines or quarries</Checkboxes.Box>
          <Checkboxes.Box value="farm">Farm or agricultural waste</Checkboxes.Box>
          <Checkboxes.Divider />
          <Checkboxes.Box value="none" id="none" exclusive>
            None
          </Checkboxes.Box>
        </Checkboxes>
      </form>,
    );

    expect(container).toMatchSnapshot();
  });

  it('Sets a data-exclusive attribute when exclusive is true for a box', () => {
    const { container } = render(
      <form>
        <Checkboxes id="example" name="example">
          <Checkboxes.Box value="animal">Waste from animal carcasses</Checkboxes.Box>
          <Checkboxes.Box value="mines">Waste from mines or quarries</Checkboxes.Box>
          <Checkboxes.Box value="farm">Farm or agricultural waste</Checkboxes.Box>
          <Checkboxes.Box value="none" id="none" exclusive>
            None
          </Checkboxes.Box>
        </Checkboxes>
      </form>,
    );

    expect(container.querySelector('#none')?.getAttribute('data-checkbox-exclusive')).toBe('true');
  });
});
