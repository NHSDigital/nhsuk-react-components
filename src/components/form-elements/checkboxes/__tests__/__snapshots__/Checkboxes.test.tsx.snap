// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkboxes Matches the snapshot with an exclusive checkbox 1`] = `
<div>
  <form>
    <div
      class="nhsuk-form-group"
    >
      <div
        class="nhsuk-checkboxes"
        id="example"
      >
        <div
          class="nhsuk-checkboxes__item"
        >
          <input
            class="nhsuk-checkboxes__input"
            data-checkbox-exclusive-group="example"
            id="example-1"
            name="example"
            type="checkbox"
            value="animal"
          />
          <label
            class="nhsuk-label nhsuk-checkboxes__label"
            for="example-1"
            id="example-1--label"
          >
            Waste from animal carcasses
          </label>
        </div>
        <div
          class="nhsuk-checkboxes__item"
        >
          <input
            class="nhsuk-checkboxes__input"
            data-checkbox-exclusive-group="example"
            id="example-2"
            name="example"
            type="checkbox"
            value="mines"
          />
          <label
            class="nhsuk-label nhsuk-checkboxes__label"
            for="example-2"
            id="example-2--label"
          >
            Waste from mines or quarries
          </label>
        </div>
        <div
          class="nhsuk-checkboxes__item"
        >
          <input
            class="nhsuk-checkboxes__input"
            data-checkbox-exclusive-group="example"
            id="example-3"
            name="example"
            type="checkbox"
            value="farm"
          />
          <label
            class="nhsuk-label nhsuk-checkboxes__label"
            for="example-3"
            id="example-3--label"
          >
            Farm or agricultural waste
          </label>
        </div>
        <div
          class="nhsuk-checkboxes__divider"
        >
          or
        </div>
        <div
          class="nhsuk-checkboxes__item"
        >
          <input
            class="nhsuk-checkboxes__input"
            data-checkbox-exclusive="true"
            data-checkbox-exclusive-group="example"
            id="none"
            name="example"
            type="checkbox"
            value="none"
          />
          <label
            class="nhsuk-label nhsuk-checkboxes__label"
            for="none"
            id="none--label"
          >
            None
          </label>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`Checkboxes matches snapshot 1`] = `
<div>
  <form>
    <div
      class="nhsuk-form-group"
    >
      <div
        class="nhsuk-checkboxes"
        id="example"
      >
        <div
          class="nhsuk-checkboxes__item"
        >
          <input
            class="nhsuk-checkboxes__input"
            data-checkbox-exclusive-group="example"
            id="example-1"
            name="example"
            type="checkbox"
            value="animal"
          />
          <label
            class="nhsuk-label nhsuk-checkboxes__label"
            for="example-1"
            id="example-1--label"
          >
            Waste from animal carcasses
          </label>
        </div>
        <div
          class="nhsuk-checkboxes__item"
        >
          <input
            class="nhsuk-checkboxes__input"
            data-checkbox-exclusive-group="example"
            id="example-2"
            name="example"
            type="checkbox"
            value="mines"
          />
          <label
            class="nhsuk-label nhsuk-checkboxes__label"
            for="example-2"
            id="example-2--label"
          >
            Waste from mines or quarries
          </label>
        </div>
        <div
          class="nhsuk-checkboxes__item"
        >
          <input
            class="nhsuk-checkboxes__input"
            data-checkbox-exclusive-group="example"
            id="example-3"
            name="example"
            type="checkbox"
            value="farm"
          />
          <label
            class="nhsuk-label nhsuk-checkboxes__label"
            for="example-3"
            id="example-3--label"
          >
            Farm or agricultural waste
          </label>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`Checkboxes matches snapshot with boolean error 1`] = `
<div>
  <form>
    <div
      class="nhsuk-form-group nhsuk-form-group--error"
    >
      <div
        class="nhsuk-checkboxes"
        id="example"
      >
        <div
          class="nhsuk-checkboxes__item"
        >
          <input
            class="nhsuk-checkboxes__input"
            data-checkbox-exclusive-group="example"
            id="example-1"
            name="example"
            type="checkbox"
            value="animal"
          />
          <label
            class="nhsuk-label nhsuk-checkboxes__label"
            for="example-1"
            id="example-1--label"
          >
            Waste from animal carcasses
          </label>
        </div>
        <div
          class="nhsuk-checkboxes__item"
        >
          <input
            class="nhsuk-checkboxes__input"
            data-checkbox-exclusive-group="example"
            id="example-2"
            name="example"
            type="checkbox"
            value="mines"
          />
          <label
            class="nhsuk-label nhsuk-checkboxes__label"
            for="example-2"
            id="example-2--label"
          >
            Waste from mines or quarries
          </label>
        </div>
        <div
          class="nhsuk-checkboxes__item"
        >
          <input
            class="nhsuk-checkboxes__input"
            data-checkbox-exclusive-group="example"
            id="example-3"
            name="example"
            type="checkbox"
            value="farm"
          />
          <label
            class="nhsuk-label nhsuk-checkboxes__label"
            for="example-3"
            id="example-3--label"
          >
            Farm or agricultural waste
          </label>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`Checkboxes matches snapshot with string error 1`] = `
<div>
  <form>
    <div
      class="nhsuk-form-group nhsuk-form-group--error"
    >
      <span
        class="nhsuk-error-message"
        id="example--error-message"
      >
        <span
          class="nhsuk-u-visually-hidden"
        >
          Error: 
        </span>
        Example error
      </span>
      <div
        class="nhsuk-checkboxes"
        id="example"
      >
        <div
          class="nhsuk-checkboxes__item"
        >
          <input
            class="nhsuk-checkboxes__input"
            data-checkbox-exclusive-group="example"
            id="example-1"
            name="example"
            type="checkbox"
            value="animal"
          />
          <label
            class="nhsuk-label nhsuk-checkboxes__label"
            for="example-1"
            id="example-1--label"
          >
            Waste from animal carcasses
          </label>
        </div>
        <div
          class="nhsuk-checkboxes__item"
        >
          <input
            class="nhsuk-checkboxes__input"
            data-checkbox-exclusive-group="example"
            id="example-2"
            name="example"
            type="checkbox"
            value="mines"
          />
          <label
            class="nhsuk-label nhsuk-checkboxes__label"
            for="example-2"
            id="example-2--label"
          >
            Waste from mines or quarries
          </label>
        </div>
        <div
          class="nhsuk-checkboxes__item"
        >
          <input
            class="nhsuk-checkboxes__input"
            data-checkbox-exclusive-group="example"
            id="example-3"
            name="example"
            type="checkbox"
            value="farm"
          />
          <label
            class="nhsuk-label nhsuk-checkboxes__label"
            for="example-3"
            id="example-3--label"
          >
            Farm or agricultural waste
          </label>
        </div>
      </div>
    </div>
  </form>
</div>
`;
